File: backend-springboot/src/main/java/com/anime/backend/controller/AnimesController.java
package com.anime.backend.controller;

import com.anime.backend.model.Anime;
import com.anime.backend.repository.AnimeRepository;
import com.anime.backend.service.AnimeImportService;
import com.anime.backend.service.ImportState;
import jakarta.persistence.criteria.Predicate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import java.time.LocalDate;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/animes")
@CrossOrigin(origins = "*") // Allow all origins like Django
public class AnimesController {

    @Autowired
    private AnimeRepository animeRepository;

    @Autowired
    private AnimeImportService animeImportService;

    @Autowired
    private ImportState importState;

    @GetMapping({"", "/"})
    public ResponseEntity<?> getAnimes(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) Double min_score,
            @RequestParam(defaultValue = "-updated_at") String ordering) {

        int pageSize = 20;

        Specification<Anime> spec = (root, query, cb) -> {
            List<Predicate> predicates = new ArrayList<>();
            if (search != null && !search.isEmpty()) {
                predicates.add(cb.like(cb.lower(root.get("title")), "%" + search.toLowerCase() + "%"));
            }
            if (category != null && !category.isEmpty()) {
                predicates.add(cb.equal(root.get("category"), category));
            }
            if (type != null && !type.isEmpty()) {
                predicates.add(cb.equal(root.get("type"), type));
            }
            if (min_score != null) {
                predicates.add(cb.greaterThanOrEqualTo(root.get("score"), min_score));
            }
            return cb.and(predicates.toArray(new Predicate[0]));
        };

        Sort sort = Sort.unsorted();
        if (ordering != null && !ordering.isEmpty()) {
            boolean desc = ordering.startsWith("-");
            String field = ordering.replaceFirst("^-", "");
            if (field.equals("updated_at")) field = "updatedAt";
            if (field.equals("personal_rating")) field = "personalRating";
            
            sort = desc ? Sort.by(field).descending() : Sort.by(field).ascending();
        }

        Pageable pageable = PageRequest.of(page - 1, pageSize, sort);
        Page<Anime> animePage = animeRepository.findAll(spec, pageable);

        String next = animePage.hasNext() ? "?page=" + (page + 1) : null;
        String previous = animePage.hasPrevious() ? "?page=" + (page - 1) : null;

        Map<String, Object> response = new java.util.LinkedHashMap<>();
        response.put("count", animePage.getTotalElements());
        response.put("next", next);
        response.put("previous", previous);
        response.put("results", animePage.getContent());

        return ResponseEntity.ok(response);
    }

    @PostMapping({"", "/"})
    public ResponseEntity<Anime> createAnime(@RequestBody Anime anime) {
        Anime saved = animeRepository.save(anime);
        return ResponseEntity.status(HttpStatus.CREATED).body(saved);
    }

    @GetMapping({"/{id}", "/{id}/"})
    public ResponseEntity<Anime> getAnime(@PathVariable Long id) {
        Optional<Anime> anime = animeRepository.findById(id);
        return anime.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PutMapping({"/{id}", "/{id}/"})
    public ResponseEntity<Anime> updateAnime(@PathVariable Long id, @RequestBody Anime animeDetails) {
        return animeRepository.findById(id).map(anime -> {
            anime.setMalId(animeDetails.getMalId());
            anime.setTitle(animeDetails.getTitle());
            anime.setUrl(animeDetails.getUrl());
            anime.setCategory(animeDetails.getCategory());
            anime.setImageUrl(animeDetails.getImageUrl());
            anime.setImageBase64(animeDetails.getImageBase64());
            anime.setScore(animeDetails.getScore());
            anime.setYear(animeDetails.getYear());
            anime.setType(animeDetails.getType());
            anime.setSynopsis(animeDetails.getSynopsis());
            anime.setPersonalRating(animeDetails.getPersonalRating());
            Anime updated = animeRepository.save(anime);
            return ResponseEntity.ok(updated);
        }).orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PatchMapping({"/{id}", "/{id}/"})
    public ResponseEntity<Anime> patchAnime(@PathVariable Long id, @RequestBody Map<String, Object> updates) {
        return animeRepository.findById(id).map(anime -> {
            if (updates.containsKey("category")) {
                anime.setCategory((String) updates.get("category"));
            }
            if (updates.containsKey("personal_rating")) {
                Object rating = updates.get("personal_rating");
                if (rating == null) {
                    anime.setPersonalRating(null);
                } else if (rating instanceof Number) {
                    anime.setPersonalRating(((Number) rating).doubleValue());
                }
            }
            if (updates.containsKey("watched_date")) {
                Object date = updates.get("watched_date");
                if (date == null || ((String) date).isEmpty()) {
                    anime.setWatchedDate(null);
                } else {
                    anime.setWatchedDate(LocalDate.parse((String) date));
                }
            }
            Anime updated = animeRepository.save(anime);
            return ResponseEntity.ok(updated);
        }).orElseGet(() -> ResponseEntity.notFound().build());
    }

    @DeleteMapping({"/{id}", "/{id}/"})
    public ResponseEntity<?> deleteAnime(@PathVariable Long id) {
        return animeRepository.findById(id).map(anime -> {
            animeRepository.delete(anime);
            return ResponseEntity.ok(Map.of("status", "success"));
        }).orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PostMapping({"/upload_list", "/upload_list/"})
    public ResponseEntity<?> uploadList(@RequestParam("file") MultipartFile file) {
        if (file.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("error", "No file uploaded."));
        }
        try {
            String content = new String(file.getBytes(), StandardCharsets.UTF_8);
            animeImportService.startProcessing(content);
            return ResponseEntity.ok(Map.of("message", "File uploaded successfully. Processing in background..."));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    public static class MalAddRequest {
        public int mal_id;
        public String title;
        public String url;
        public String category = "Plan to watch";
        public String image_url;
        public Double score;
        public Integer year;
        public String type;
        public String synopsis;
    }

    @PostMapping({"/add_from_mal", "/add_from_mal/"})
    public ResponseEntity<?> addFromMal(@RequestBody MalAddRequest req) {
        String title = req.title != null ? req.title : "Unknown Title";
        String url = req.url != null ? req.url : "https://myanimelist.net/anime/" + req.mal_id;

        animeImportService.upsertAnime(req.mal_id, title, url, req.category, null);
        Optional<Anime> anime = animeRepository.findByMalId(req.mal_id);
        
        return anime.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build());
    }

    @GetMapping({"/import_status", "/import_status/"})
    public ResponseEntity<?> importStatus() {
        return ResponseEntity.ok(importState.getStateSnapshot());
    }
}

File: backend-dotnetWebApi/Controllers/AnimesController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using backend_dotnetWebApi.Data;
using backend_dotnetWebApi.Models;
using backend_dotnetWebApi.Services;

namespace backend_dotnetWebApi.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class AnimesController : ControllerBase
    {
        private readonly AnimeContext _context;
        private readonly ImportState _importState;
        private readonly AnimeImportService _importService;

        public AnimesController(AnimeContext context, ImportState importState, AnimeImportService importService)
        {
            _context = context;
            _importState = importState;
            _importService = importService;
        }

        [HttpGet]
        public async Task<ActionResult<object>> GetAnimes(
            [FromQuery] int page = 1,
            [FromQuery] string? search = null,
            [FromQuery] string? category = null,
            [FromQuery] string? type = null,
            [FromQuery] double? min_score = null,
            [FromQuery] string? ordering = "-updated_at")
        {
            int pageSize = 20;
            var query = _context.Animes.AsQueryable();

            if (!string.IsNullOrEmpty(search))
            {
                query = query.Where(a => EF.Functions.Like(a.Title, $"%{search}%"));
            }
            if (!string.IsNullOrEmpty(category))
            {
                query = query.Where(a => a.Category == category);
            }
            if (!string.IsNullOrEmpty(type))
            {
                query = query.Where(a => a.Type == type);
            }
            if (min_score.HasValue)
            {
                query = query.Where(a => a.Score >= min_score.Value);
            }

            // Ordering
            if (!string.IsNullOrEmpty(ordering))
            {
                bool desc = ordering.StartsWith("-");
                string field = ordering.TrimStart('-');
                
                query = field.ToLower() switch
                {
                    "updated_at" => desc ? query.OrderByDescending(a => a.UpdatedAt) : query.OrderBy(a => a.UpdatedAt),
                    "score" => desc ? query.OrderByDescending(a => a.Score) : query.OrderBy(a => a.Score),
                    "year" => desc ? query.OrderByDescending(a => a.Year) : query.OrderBy(a => a.Year),
                    "title" => desc ? query.OrderByDescending(a => a.Title) : query.OrderBy(a => a.Title),
                    "personal_rating" => desc ? query.OrderByDescending(a => a.PersonalRating) : query.OrderBy(a => a.PersonalRating),
                    _ => query.OrderByDescending(a => a.UpdatedAt)
                };
            }

            int total = await query.CountAsync();
            var items = await query.Skip((page - 1) * pageSize).Take(pageSize).ToListAsync();

            string? next = (page * pageSize) < total ? $"?page={page + 1}" : null;
            string? prev = page > 1 ? $"?page={page - 1}" : null;

            return new
            {
                count = total,
                next = next,
                previous = prev,
                results = items
            };
        }

        [HttpGet("{id}")]
        public async Task<ActionResult<Anime>> GetAnime(long id)
        {
            var anime = await _context.Animes.FindAsync(id);
            if (anime == null) return NotFound();
            return anime;
        }

        [HttpPost]
        public async Task<ActionResult<Anime>> PostAnime(Anime anime)
        {
            _context.Animes.Add(anime);
            await _context.SaveChangesAsync();
            return CreatedAtAction(nameof(GetAnime), new { id = anime.Id }, anime);
        }

        [HttpPut("{id}")]
        public async Task<IActionResult> PutAnime(long id, Anime anime)
        {
            if (id != anime.Id) return BadRequest();
            _context.Entry(anime).State = EntityState.Modified;
            await _context.SaveChangesAsync();
            return NoContent();
        }

        [HttpPatch("{id}")]
        public async Task<IActionResult> PatchAnime(long id, [FromBody] System.Text.Json.Nodes.JsonObject updates)
        {
            var anime = await _context.Animes.FindAsync(id);
            if (anime == null) return NotFound();

            if (updates.TryGetPropertyValue("category", out var categoryNode) && categoryNode != null)
            {
                anime.Category = categoryNode.ToString();
            }
            if (updates.TryGetPropertyValue("personal_rating", out var ratingNode))
            {
                anime.PersonalRating = ratingNode == null ? null : (double?)ratingNode.AsValue();
            }
            if (updates.TryGetPropertyValue("watched_date", out var dateNode))
            {
                if (dateNode == null)
                {
                    anime.WatchedDate = null;
                }
                else
                {
                    string dateStr = dateNode.ToString();
                    anime.WatchedDate = string.IsNullOrEmpty(dateStr) ? null : DateTime.Parse(dateStr);
                }
            }

            anime.UpdatedAt = DateTimeOffset.UtcNow;
            await _context.SaveChangesAsync();
            return Ok(anime);
        }

        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteAnime(long id)
        {
            var anime = await _context.Animes.FindAsync(id);
            if (anime == null) return NotFound();
            _context.Animes.Remove(anime);
            await _context.SaveChangesAsync();
            return Ok(new { status = "success" });
        }

        [HttpPost("upload_list")]
        public async Task<IActionResult> UploadList(IFormFile file)
        {
            if (file == null || file.Length == 0)
                return BadRequest(new { error = "No file uploaded." });

            using var reader = new StreamReader(file.OpenReadStream(), Encoding.UTF8);
            var content = await reader.ReadToEndAsync();
            
            _importService.StartProcessing(content);
            return Ok(new { message = "File uploaded successfully. Processing in background..." });
        }

        public class MalAddRequest
        {
            public int mal_id { get; set; }
            public string? title { get; set; }
            public string? url { get; set; }
            public string? category { get; set; } = "Plan to watch";
            public string? image_url { get; set; }
            public double? score { get; set; }
            public int? year { get; set; }
            public string? type { get; set; }
            public string? synopsis { get; set; }
        }

        [HttpPost("add_from_mal")]
        public async Task<IActionResult> AddFromMal([FromBody] MalAddRequest req)
        {
            string url = req.url ?? $"https://myanimelist.net/anime/{req.mal_id}";
            string title = req.title ?? "Unknown Title";

            await _importService.UpsertAnimeAsync(req.mal_id, title, url, req.category, null);
            var anime = await _context.Animes.FirstOrDefaultAsync(a => a.MalId == req.mal_id);
            return Ok(anime);
        }

        [HttpGet("import_status")]
        public IActionResult ImportStatus()
        {
            return Ok(_importState.GetStateSnapshot());
        }
    }
}
